support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -1,12 +1,79 @@
 package com.gu.support.workers
 
+import cats.implicits._
 import com.gu.i18n.Title
 import com.gu.support.encoding.Codec
 import com.gu.support.encoding.Codec.deriveCodec
 import com.gu.support.encoding.CustomCodecs._
+import com.gu.support.workers.GiftRecipient.{DigiSub, Weekly}
+import io.circe._
+import io.circe.syntax._
 
-case class GiftRecipient(title: Option[Title], firstName: String, lastName: String, email: Option[String])
-
+sealed trait GiftRecipient {
+  val giftRecipientType: String
+  val firstName: String
+  val lastName: String
+  def asDigiSub: Option[DigiSub]
+  def asWeekly: Option[Weekly]
+}
 object GiftRecipient {
-  implicit val codec: Codec[GiftRecipient] = deriveCodec
+
+  implicit val c1: Codec[Weekly] = deriveCodec
+  implicit val c2: Codec[DigiSub] = deriveCodec
+
+  implicit val encoder: Encoder[GiftRecipient] = Encoder.instance {
+    case c: Weekly => c.asJson
+    case c: DigiSub => c.asJson
+  }
+
+  def or[A, AA >: A](da: => Decoder[A], daa: => Decoder[AA]): Decoder[AA] = new Decoder[AA] {
+    final def apply(c: HCursor): Decoder.Result[AA] = da(c) match {
+      case r @ Right(_) => r
+      case Left(failure)      => daa(c).leftMap(secondFail => DecodingFailure(failure.message + " / " + secondFail.message, failure.history ++ secondFail.history))
+    }
+  }
+
+  implicit val decoder: Decoder[GiftRecipient] =
+    or(Decoder[Weekly], Decoder[DigiSub].widen)
+
+  case class Weekly(
```

Comment:
Still thinking about the serialisation approach but can we call these classes `WeeklyGiftRecipient` and `DigitalGiftRecipient` so they more specifically describe what they represent? I would actually probably go for `GuardianWeeklyGiftRecipient` and DigitalSusbcriptionGiftRecipient` but that's a bit of a mouthful!

Reply:
thanks rupert will do on the longer names.

In the slightly longer term I'd like to add a convention to our coding standards about ADTs structure as we have a few different ways
e.g. 
- one side we have several Monthly in the codebase including `com.gu.support.workers.Monthly` (which is actually a BillingPeriod) on one hand and
- the other side we have things like `com.gu.support.workers.GiftRecipient.WeeklyGiftRecipient` which gives great auto complete from entering `GiftRecipient.` but has a bit of duplication if you are looking at the whole type name.

I actually think/agree the latter is best overall as it's both clear in github, and gives good auto complete, but if we write it down as a rule then it becomes easier to be consistent when adding and refactoring.